core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -3843,43 +3843,44 @@ object Stream extends StreamLowPriority {
         // "block" and await until the `running` counter drops to zero.
         val awaitWhileRunning: F[Unit] = running.discrete.dropWhile(_ > 0).take(1).compile.drain
 
-        // runs one inner stream,
-        // each stream is forked.
+        // signals that a running stream, either inner or ourer, finished with or without failure.
+        def endWithResult(result: Either[Throwable, Unit]): F[Unit] =
+          result match {
+            case Right(()) => decrementRunning
+            case Left(err) => stop(Some(err)) >> decrementRunning
+          }
+
+        def leaseOrFail(scope: Scope[F]): F[Scope.Lease[F]] =
```

Comment:
Thoughts on moving this directly to `Scope`? I just ran in to the need for the same thing in #1982.